File: volatility-service/src/main/kotlin/com/kinetix/volatility/interp/SurfaceInterp.kt
package com.kinetix.volatility.interp

/**
 * Bilinear vol-surface interpolation with flat extrapolation.
 *
 * Mirrors the Python risk-engine helper but lives on the Kotlin side
 * so the volatility-service can serve interpolated vols without a
 * gRPC round-trip. Inputs are strike + expiry knot arrays + a 2D vol
 * grid; queries return the interpolated vol at any (strike, expiry).
 */
data class VolSurfaceGrid(
    val strikes: List<Double>,
    val expiryDays: List<Int>,
    val vols: List<List<Double>>,
) {
    init {
        require(strikes.isNotEmpty() && expiryDays.isNotEmpty()) {
            "VolSurfaceGrid needs at least one strike and one expiry"
        }
        require(vols.size == strikes.size) {
            "vols has ${vols.size} rows but strikes has ${strikes.size}"
        }
        for (row in vols) {
            require(row.size == expiryDays.size) {
                "vols row length must equal expiryDays size"
            }
        }
    }
}

fun interpolateSurfaceVol(grid: VolSurfaceGrid, strike: Double, expiryDays: Int): Double {
    val (iLo, iHi, wi) = bracket(grid.strikes, strike)
    val (jLo, jHi, wj) = bracket(grid.expiryDays.map { it.toDouble() }, expiryDays.toDouble())
    val v00 = grid.vols[iLo][jLo]
    val v01 = grid.vols[iLo][jHi]
    val v10 = grid.vols[iHi][jLo]
    val v11 = grid.vols[iHi][jHi]
    return (1 - wi) * (1 - wj) * v00 +
        (1 - wi) * wj * v01 +
        wi * (1 - wj) * v10 +
        wi * wj * v11
}

private fun bracket(axis: List<Double>, value: Double): Triple<Int, Int, Double> {
    if (value <= axis.first()) return Triple(0, 0, 0.0)
    if (value >= axis.last()) return Triple(axis.size - 1, axis.size - 1, 0.0)
    for (i in 0 until axis.size - 1) {
        if (axis[i] <= value && value <= axis[i + 1]) {
            val span = axis[i + 1] - axis[i]
            if (span == 0.0) return Triple(i, i + 1, 0.0)
            return Triple(i, i + 1, (value - axis[i]) / span)
        }
    }
    return Triple(axis.size - 1, axis.size - 1, 0.0)
}

File: risk-engine/src/kinetix_risk/surface_interp.py
"""Bivariate volatility-surface interpolator.

Implied-vol surfaces are sampled on a grid of (strike, expiry) knots
but pricers need vol at arbitrary points. Bilinear interpolation
(2D linear) gives a smooth-enough surface for risk calc and avoids
the cubic-spline pitfall of negative vols in extrapolated corners.

References
----------
Gatheral, J. (2006). *The Volatility Surface*. Wiley. Chapter 1
    covers grid-based surface representation; the practitioner
    interpolation choices balance smoothness against the risk of
    introducing arbitrage or negative-vol points.
"""

from dataclasses import dataclass


@dataclass(frozen=True)
class VolSurfaceGrid:
    """Implied-vol grid: strikes x expiries x vols."""

    strikes: tuple[float, ...]   # ascending
    expiries_days: tuple[int, ...]   # ascending
    vols: tuple[tuple[float, ...], ...]   # vols[i_strike][j_expiry]

    def __post_init__(self) -> None:
        if not self.strikes or not self.expiries_days:
            raise ValueError("VolSurfaceGrid needs at least one strike and one expiry")
        if len(self.vols) != len(self.strikes):
            raise ValueError(
                f"vols has {len(self.vols)} rows but strikes has {len(self.strikes)}",
            )
        for row in self.vols:
            if len(row) != len(self.expiries_days):
                raise ValueError("vols row length must equal len(expiries_days)")


def interpolate_vol(grid: VolSurfaceGrid, strike: float, expiry_days: int) -> float:
    """Bilinear interpolation of vol at (strike, expiry_days).

    Outside the knot range, clamps to the boundary (flat
    extrapolation) — same convention as the rates interpolators. The
    flat-extrapolation choice avoids the cubic-spline failure mode
    where a deeply OTM strike picks up a negative vol from the
    extrapolated tail.
    """
    strikes = grid.strikes
    expiries = grid.expiries_days
    i_lo, i_hi, wi = _bracket(strikes, strike)
    j_lo, j_hi, wj = _bracket(expiries, expiry_days)
    v00 = grid.vols[i_lo][j_lo]
    v01 = grid.vols[i_lo][j_hi]
    v10 = grid.vols[i_hi][j_lo]
    v11 = grid.vols[i_hi][j_hi]
    return (
        (1 - wi) * (1 - wj) * v00
        + (1 - wi) * wj * v01
        + wi * (1 - wj) * v10
        + wi * wj * v11
    )


def _bracket(axis: tuple, value: float) -> tuple[int, int, float]:
    """Return (low_index, high_index, weight) such that the value sits
    at ``low_index + weight`` along the axis (linear interpolation
    parameter in [0, 1])."""
    if value <= axis[0]:
        return 0, 0, 0.0
    if value >= axis[-1]:
        return len(axis) - 1, len(axis) - 1, 0.0
    for i in range(len(axis) - 1):
        if axis[i] <= value <= axis[i + 1]:
            span = axis[i + 1] - axis[i]
            if span == 0:
                return i, i + 1, 0.0
            return i, i + 1, (value - axis[i]) / span
    return len(axis) - 1, len(axis) - 1, 0.0
